libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/GenState.hs

```diff
@@ -202,28 +360,82 @@ genDatumWithHash = do
     ts {gsDatums = Map.insert datumHash datum gsDatums}
   pure (datumHash, datum)
 
--- Adds to the gsDPState
+-- Adds to the rewards of the ModelNewEpochState
 genRewards :: Reflect era => GenRS era (RewardAccounts (Crypto era))
 genRewards = do
-  n <- lift nonNeg1Digit
-  newrewards <-
-    Map.fromList <$> replicateM n ((,) <$> genCredential Rewrd <*> lift genPositiveVal)
-  modifyDState $ \ds -> ds {_unified = rewards ds UM.⨃ newrewards} -- Prefers coins in newrewards
+  wmax <- gets (withdrawalMax . geSize . gsGenEnv)
+  n <- lift $ choose (1, wmax)
+  old <- gets (Map.keysSet . gsInitialRewards) -- we need a fresh credential, one that was not previously
+  prev <- gets gsRegKey -- generated here, or one that arose from RegKey (i.e. prev)
+  credentials <- genFreshCredentials n 100 Rewrd (Set.union old prev) []
+  -- newrewards <-
+  -- Map.fromList <$> replicateM n ((,) <$> genCredential Rewrd <*> lift genRewardVal)
+  -- always generate a New Reward.
+  -- Map.fromList <$> replicateM n ((,) <$> frequencyT [ (1, KeyHashObj <$> genKeyHash),(35, ScriptHashObj <$> genScript reify Rewrd)] <*> lift genRewardVal)
+  newrewards <- Map.fromList <$> mapM (\x -> (,) x <$> lift genRewardVal) credentials
+  modifyModel (\m -> m {mRewards = eval (mRewards m ⨃ newrewards)}) -- Prefers coins in newrewards
   pure newrewards
 
--- Adds to PState part of DPSTate
-genPool :: Reflect era => GenRS era (KeyHash 'StakePool (Crypto era))
+-- | Generate a 'n' fresh credentials (ones not in the set 'old'). We get 'trys' chances,
+--   if it doesn't work in 'trys' attempts then quit with an error. Better to raise an error
+--   than go into an infinite loop.
+genFreshCredentials ::
+  forall era kr.
+  Reflect era =>
+  Int ->
+  Int ->
+  Tag ->
+  Set (Credential kr (Crypto era)) ->
+  [Credential kr (Crypto era)] ->
+  GenRS era [Credential kr (Crypto era)]
+genFreshCredentials 0 _trys _tag _old ans = pure ans
+genFreshCredentials _n 0 _tag _old _ans = error ("Ran out of trys in genFreshCredentials.")
+genFreshCredentials n trys tag old ans = do
+  c <- genFreshCredential trys tag old
+  genFreshCredentials (n - 1) trys tag (Set.insert c old) (c : ans)
```

Comment:
his is a general suggestion: there is overhead for passing arguments that don't change in a recursive function. It is better to create a local binding that captures arguments that don't change, eg:
```suggestion
genFreshCredentials n trys tag 
  | trys == 0 = error "Invalid value for number of trys"
  | otherwise = go n
  where 
    go i old ans
      | i == 0 = pure ans
      | otheriwise = do
          c <- genFreshCredential trys tag old
          go (i - 1) trys tag (Set.insert c old) (c : ans)
```

Side question: shouldn't it be spelled "tries" and not "trys"?

Reply:
I made this change, but I think it is a bit ridiculous. I'll bet this function never iterates more than 2 or 3 times. I'd hardly call this an inner loop, where efficiency matters.